VerilogForVerificationStyle.md

```diff
@@ -0,0 +1,826 @@
+
+
+### Enums
+
+Enum names should be written in `lower_snake_case`, and the enumerated values
+should be written in `UpperCamelCase` as follows:
+
+```systemverilog
+// Enum type names should include the block name for clarity.
+// Enum value names should include the name of the enum.
+typedef enum bit {
+  UartInterruptFrameErr,
+  UartInterruptRxBufferFull,
+  …
+} uart_interrupte_e;
+```
+
+
+### Interfaces, Clocking Blocks, Modports
+
+1. Do not use the `program` construct.
+2. Clocks and reset must be generated in SystemVerilog modules, interfaces,
+or in the UVM component hierarchy.
+Do not do this generation in `uvm_object` derived classes.
+3. Use clocking blocks in a SystemVerilog interface to sample and drive
+synchronous DUT interfaces.
+4. It is recommended to implement a reusable interface scheme for block level
+testbenches that will be included in integration/system level testbenches.
+
+
+### Loop Operators
+
+It is highly recommended to use `foreach` loops as often as possible for
+iteration functionality, unless absolutely necessary.
+They are concise and less prone to errors.
+
+
+### Code Within Asserts
+
+For any randomization that is required, you must use the macros specified
+in the [Randomization Macros section](#macro-usage), do not use `assert()` to check
+manual randomization calls, use the provided macros instead.
+More broadly, do not place any expression with side effects into an
+`assert()` statement.
+
+:-1:
+```systemverilog
+// Do not assert the output of randomize() calls
+// 1.
+ret = item.randomize() with {<constraints>};
+assert(ret);
+// 2.
+assert(randomize(<variable>) with {<constraints>});
+// 3.
+assert(item.randomize() with {<constraints>});
+// 4.
+assert(function_that_has_side_effects());
+```
+
+### Wait and Fork
+
+Always put wait fork and disable fork constructs inside of an
+isolation fork...join block to avoid erroneous waiting.
+It is recommended to use the `DV_SPINWAIT` macro whenever possible,
+[as described here](#macro-usage)
+
+
+### Void Casts
+
+Do not void cast any function/task calls that have useful return values,
+EXCEPT the following:
+* System functions
+* RAL `predict()` calls
+* FIFO/Queue methods
+
+
+### Associative Arrays
+
+Do not use wildcard indexed [*] associative arrays.
+Always specify a particular index type.
+
+
+### Bind Statements
+
+`bind` statements are the preferred approach for using assertion-based monitors.
+For these situations, using `.*` is allowed for making implicit port
+declarations when binding a module whose ports are all inputs.
+
+
+### Simulator-specific Code
+
+Always wrap simulator-specific code inside preprocessor guards.
+The macro names `VCS`, `INCA`, `XCELIUM` are defined for the VCS,
+Incisive/NCsim/Irun, and Xcelium/xrun simulators respectively,
+these must be used to wrap relevant code, as shown below:
+
+```systemverilog
+`ifdef VCS
+  $stack();
+`endif
+```
+
+
+### Forbidden System Tasks and Functions
```

Comment:
Maybe use "discouraged" instead of "forbidden"?

Reply:
I feel that 'forbidden' would be slightly better for the purpose of this section, as the listed system functions must never be used in any lowRISC DV environment, as opposed to providing a recommendation to not use the functions. What do you think?

Reply:
In the design style guide we preferably use "discouraged". But I think there is no single true answer here. I guess "forbidden" is a very hard term. Question: Would you completely block the adoption/integration of code into a project if any of these functions are used? If yes, "forbidden" is the right term. If instead you would consider a step-wise process of integration first and at a later point rework the problematic parts, I would prefer "discouraged".

Anyway, I don't really mind. Please feel free to take the decision.

Reply:
'Embargoed'?

Reply:
IMO, use of these functions should block adoption of code into a project, as these functions are either not in the SystemVerilog LRM or are not part of the SV randomization stability model (which could break reproducibility). I will add this rationale to the document.
What do you think @sriyerg?

Reply:
I think forbidden can be justified in some cases, in particular the `$random` and `$dist_*` mentioned below which can break simulation reproducibility. Breaking reproducibility can be very painful and hard to diagnose so this seems a good reason to forbid those functions.

The others don't seem to cause such a serious problem if they occur but as they're not part of the LRM could break the build if you have a simulator that doesn't support them, again a decent reason to forbid them.

Reply:
Agree - we may have much control over 3rd party code, but if they are using these, then we file bugs and get them to fix it. We do want to forbid the active contributors of OpenTitan from using these. $psprintf is problematic too since its not a part of the LRM and some simulators may not support it.

Reply:
Agree. IMO, it's good to forbid these items.